add rspec-run-last-failed to rerun only failed examples

ledbettj opened the pull request:
rspec-run-last-failed will re-run the specs that failed during the last invocation.  By default this is bound to `C-c ,f`.

I'm not an elisp pro so I'm sure there's some cleanup that can be done here.  Feedback welcome!

rranelli commented:
I have not tested it yet, but I think this is an awesome feature :+1:

dgutov commented on rspec-mode.el:
```diff
@@ -668,6 +676,10 @@ or a cons (FILE . LINE), to run one example."
     (rspec-compile (rspec-runner-target spec-files) opts)))
 
+(defvar rspec-last-failed-specs
+  '()
```
`'()` is the same as nil, and no need to put it on a separate line.

dgutov commented on rspec-mode.el:
```diff
@@ -718,8 +730,24 @@ or a cons (FILE . LINE), to run one example."
 (define-compilation-mode rspec-compilation-mode "RSpec Compilation"
   "Compilation mode for RSpec output."
   (add-hook 'compilation-filter-hook 'rspec-colorize-compilation-buffer nil t)
+  (add-hook 'compilation-finish-functions 'rspec-store-failures)
```
Add it locally, not globally.

dgutov commented on rspec-mode.el:
```diff
@@ -718,8 +730,24 @@ or a cons (FILE . LINE), to run one example."
+      (reverse matches))))
+
+(defun rspec-store-failures (compilation-buffer process-result)
+  (let* ( (body (buffer-string))
```
Please remove the weird space.

ledbettj commented:
Changes:
* Replaced `'()` with `nil`.
* Added compilation hook locally instead of globally.
* Removed weird space.
* Fixed keybinding to be in the right keymap.
* Moved regexp logic into the rspec-store-failures function.

Thanks for the comments!

dgutov commented on rspec-mode.el:
```diff
@@ -718,8 +729,21 @@ or a cons (FILE . LINE), to run one example."
+          (failure-regexp "^rspec \\([0-9A-Za-z@_./:-]+\\.rb:[0-9]+\\)")
+          (failures (save-match-data
+                      (let ((pos 0) matches)
+                        (while (string-match failure-regexp body pos)
```
This would be better performed with `re-search-forward`. You don't copy the buffer text into a string; instead use `save-excursion`, `(goto-char (point-min))`, and traverse the buffer till the end while saving the matches,

ledbettj commented:
Thanks for the comment.  I've updated the logic, hopefully correctly.

ledbettj commented:
Dropped the `save-match-data` as well.

dgutov commented:
Squash-merged. Thanks!

ledbettj commented:
:+1: Thanks!
